### PR TITLE
Streaming responses from OpenAI and GPT4All

### PR DESCRIPTION
**Pull request by raulraja**
These changes introduce the capability of streaming in chat responses. It adds the following changes:

- A new function, `createChatCompletions`, has been added to the `Chat` interface, which allows creating chat completions based on a `ChatCompletionRequest`. This function returns a `Flow<ChatCompletionChunk>` representing the generated chat completions.
- The `Chat` interface now includes two overloaded versions of the `promptStreaming` function. These functions enable streaming by returning a `Flow<String>` that emits the generated chat responses as they become available.
- The implementation of `createChatCompletions` in `GPT4All` utilizes a `Flow` and channels to enable streaming of chat completions.
- The `MockOpenAIClient` has been updated to throw a `NotImplementedError` for `chatCompletions` since it's not implemented in the mock.
- The `OpenAIClient` implementation has been updated to utilize the OpenAI API's chat completion functionality and transform the response into the corresponding domain models (`ChatCompletionChunk`, `ChatChunk`, `ChatDelta`, etc.).

# Example
```kotlin
package com.xebia.functional.xef.auto.streaming

import com.xebia.functional.xef.auto.llm.openai.OpenAI
import com.xebia.functional.xef.auto.llm.openai.OpenAIEmbeddings
import com.xebia.functional.xef.llm.Chat
import com.xebia.functional.xef.vectorstores.LocalVectorStore

suspend fun main() {
  val chat: Chat = OpenAI.DEFAULT_CHAT
  val embeddings = OpenAIEmbeddings(OpenAI.DEFAULT_EMBEDDING)
  val vectorStore = LocalVectorStore(embeddings)
  chat.promptStreaming(
    question = "What is the meaning of life?",
    context = vectorStore
  ).collect {
    print(it)
  }
}
```

These changes enable developers to perform streaming chat completions and receive responses incrementally, enhancing the real-time interactive chat experience.

### CODE REVIEW

**Comment by raulraja** [comment]
@xebia-functional/team-ai

File: core/src/commonMain/kotlin/com/xebia/functional/xef/llm/Chat.kt [hunk 1]
```diff
@@ -9,14 +9,72 @@ import com.xebia.functional.xef.llm.models.chat.*
 import com.xebia.functional.xef.llm.models.functions.CFunction
 import com.xebia.functional.xef.prompt.Prompt
 import com.xebia.functional.xef.vectorstores.VectorStore
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
```
Comment (diesalbla) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/llm/Chat.kt:13]:
```suggestion
import kotlinx.coroutines.flow.flow
```

**Review by Yawolf — APPROVED** [review]

File: openai/src/commonMain/kotlin/com/xebia/functional/xef/auto/llm/openai/MockAIClient.kt [hunk 2]
```diff
@@ -28,6 +29,9 @@ class MockOpenAIClient(
   private val chatCompletion: (ChatCompletionRequest) -> ChatCompletionResponse = {
```
Comment (serras) [inline openai/src/commonMain/kotlin/com/xebia/functional/xef/auto/llm/openai/MockAIClient.kt:29]: Should we maybe split these methods in a different interface? I see more and more that a client implements the interface only partially.

**Review by serras — APPROVED** [review]

File: openai/src/commonMain/kotlin/com/xebia/functional/xef/auto/llm/openai/MockAIClient.kt (hunk 2 above)
Comment (raulraja) [reply openai/src/commonMain/kotlin/com/xebia/functional/xef/auto/llm/openai/MockAIClient.kt:29]: They've already been split into its own interface in `main` inside `Chat`, `ChatWithFunctions` etc. The remaining place where it implements all of them like this is the MockClient. The `AIClient` interface in the main is already unused and should be removed. I'll push a commit to this PR to remove it.

**Pull request merged**